apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/DiagnosisUtilV2.java

```diff
@@ -178,86 +177,146 @@ private static void addCodingToCodeableConcept(
    *     retrive information with.
-   * @param claim Passed as an Object because there is no top level `Claim` class that claims derive
-   *     from
+   * @param codes The mapping of diagnosis codes by their property name and respective value
+   * @param codeVersions The mapping of diagnosis code versions by their property name and
+   *     respective value
+   * @param presentOnAdms The mapping of diagnosis "PresentOnAdmissionCode" codes by their property
+   *     name and respective value
    * @param ccw CCW Codebook value that represents which "PresentOnAdmissionCode" is being used.
    *     Example: {@link CcwCodebookVariable#CLM_POA_IND_SW5}
    * @param label One or more labels to use when mapping the diagnosis.
    * @return a {@link Diagnosis} or {@link Optional#empty()}
    */
   public static Optional<Diagnosis> extractDiagnosis(
-      String substitution, Object claim, Optional<CcwCodebookInterface> ccw, DiagnosisLabel label) {
-
+      String substitution,
+      Map<String, Optional<String>> codes,
+      Map<String, Optional<Character>> codeVersions,
+      Optional<Map<String, Optional<Character>>> presentOnAdms,
+      Optional<CcwCodebookInterface> ccw,
+      DiagnosisLabel label) {
     Optional<String> code =
-        ReflectionUtils.tryMethod(claim, String.format("getDiagnosis%sCode", substitution));
+        codes.getOrDefault(String.format("diagnosis%sCode", substitution), Optional.empty());
     Optional<Character> codeVersion =
-        ReflectionUtils.tryMethod(claim, String.format("getDiagnosis%sCodeVersion", substitution));
+        codeVersions.getOrDefault(
+            String.format("diagnosis%sCodeVersion", substitution), Optional.empty());
     Optional<Character> presentOnAdm =
-        ReflectionUtils.tryMethod(
-            claim, String.format("getDiagnosis%sPresentOnAdmissionCode", substitution));
-
+        presentOnAdms.isPresent()
+            ? presentOnAdms
+                .get()
+                .getOrDefault(
+                    String.format("diagnosis%sPresentOnAdmissionCode", substitution),
+                    Optional.empty())
+            : Optional.empty();
     return Diagnosis.from(code, codeVersion, presentOnAdm, ccw, label);
   }
 
   /**
-   * Extracts nearly all diagnosis types from a Claim. It does this with reflection, so if the
-   * specific claim type doesn't have the given diagnosis it will just be skipped.
+   * Extracts nearly all diagnosis types from a Claim. If the specific claim type doesn't have the
+   * given diagnosis it will just be skipped.
    *
-   * @param claim the Claim to extract the {@link Diagnosis}es from
+   * @param codes The mapping of diagnosis codes by their property name and respective value
+   * @param codeVersions The mapping of diagnosis code versions by their property name and
+   *     respective value
+   * @param presentOnAdms The mapping of diagnosis "PresentOnAdmissionCode" codes by their property
+   *     name and respective value
```

Comment:
I noticed this is optional and only present on some claim types; is there any more information we could describe here about when this is expected or not?

Reply:
From what I gathered it's only present on `InpatientClaims` and I'd guess that's because that's the only claim where it may be relevant to know re:  this explanation in the codebase:

> Diagnosis Present On Admission (POA) Indicator: This field identifies the patient's condition that is present at the time the order for inpatient admission to a general acute care hospital occurs. This indicator is assigned to every principal and secondary diagnosis on an inpatient acute care hospital claim, and the external cause of injury codes. Conditions that develop during an outpatient encounter, including emergency department, observation, or outpatient surgery, are considered as present on admission. The Centers for Medicare and Medicaid Services (CMS) does not require a POA indicator for the external cause of injury code unless it is being reported as another diagnosis. Effective with discharges on or after 10/01/07.

Does that seem right? If so, I can add a quick comment to the method